### PR TITLE
Exercise lists with permissions

### PR DESCRIPTION
**Pull request by Dantemss**
This PR implements permissions to view unpublished exercises and solutions through lists.

- Changed how exercises work in lists so if you add them once, then all versions are in the list
- Lists that you own give list readers and list editors permission to read/edit exercises you collaborate in

### CODE REVIEW

File: db/migrate/20161020001239_move_list_exercises_and_list_vocab_terms_to_list_publication_groups.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    end
+
+    drop_table :list_exercises
+    drop_table :list_vocab_terms
+  end
+
+  def down
+    raise ActiveRecord::IrreversibleMigration
+  end
+end
```
Comment (jpslav) [inline db/migrate/20161020001239_move_list_exercises_and_list_vocab_terms_to_list_publication_groups.rb]: Are OpenStax-owned exercises already in a list?  Or will we need to make that list and move exercises into it after this PR?

File: spec/access_policies/list_publication_group_access_policy_spec.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+RSpec.describe ListPublicationGroupAccessPolicy, type: :access_policy do
+  let(:anon)        { AnonymousUser.instance }
+  let(:user)        { FactoryGirl.create(:user) }
+  let(:admin)       { FactoryGirl.create(:user, :administrator) }
+  let(:application) { FactoryGirl.create(:doorkeeper_application) }
+
+  context '' do
+    xit '' do
+    end
+  end
```
Comment (jpslav) [inline spec/access_policies/list_publication_group_access_policy_spec.rb]: :-)

File: app/models/community_solution.rb [hunk 3]
```diff
@@ -10,9 +10,6 @@ class CommunitySolution < ActiveRecord::Base
 
     joins{publication.authors.outer}
       .joins{publication.copyright_holders.outer}
-      .joins{publication.editors.outer}
-      .where{ (authors.user_id == user_id) | \
-              (copyright_holders.user_id == user_id) | \
-              (editors.user_id == user_id) }
```
Comment (Dantemss) [inline app/models/community_solution.rb]: Yep I should add the list stuff here.

File: app/models/publication.rb [hunk 4]
```diff
@@ -70,10 +69,32 @@ def is_public?
...
+    collaborators.include? user
+  end
+
+  def has_read_permission?(user)
+    has_collaborator?(user) || collaborators.any? do |collaborator|
+      collaborator.list_owners.any? do |list_owner|
+        list = list_owner.list
+        list.has_publication_group?(publication_group) && list.has_member?(user)
+      end
+    end
```
Comment (Dantemss) [inline app/models/publication.rb]: Hm I suppose I can preload the collaborator and list_owners/lists if the record is saved already...

File: spec/access_policies/list_publication_group_access_policy_spec.rb (hunk 2 above)
Comment (Dantemss) [inline spec/access_policies/list_publication_group_access_policy_spec.rb]: Not sure if I planned to add something here... maybe I'll just add the default shoulda_matchers stuff.

Comment (Dantemss) [inline spec/access_policies/list_publication_group_access_policy_spec.rb]: Oh nvm this is the access policy... yeah there should be stuff here.

File: app/access_policies/publication_access_policy.rb [hunk 5]
```diff
@@ -4,7 +4,8 @@ class PublicationAccessPolicy
   def self.action_allowed?(action, requestor, publication)
     case action
     when :read, :publish
-      !publication.is_published? && publication.has_collaborator?(requestor)
+      !publication.is_published? &&
+      publication.has_write_permission?(requestor)
```
Comment (Dantemss) [inline app/access_policies/publication_access_policy.rb]: Oh turns out this `:read` permission seems to be unused.

File: app/models/publication.rb (hunk 4 above)
Comment (Dantemss) [inline app/models/publication.rb]: OK I got rid of most of them... might still be a performance issue if we start using Accounts groups as list owners, but for now should be fine...

**Pull request merged**